### PR TITLE
Fix admin handling at merge of PR

### LINKED ISSUE #9739: Error 500 attempting to merge PR as admin
- Gitea version (or commit ref): 2ecf98ed2d10aea4f1b4ba31887aa52b479c1317

## Description

When I attempt to merge a PR that doesn't have enough approvals, the `[Merge]` button shows up in red and the message _"As an administrator, you may still merge this pull request"_ is displayed. However, when I push the `[Merge]` button, a `500 Server Error` page shows up with the following error:

```
git push: remote: Gitea: Not allowed to push to protected branch master and pr #68 is not ready to be merged: not allowed to merge [reason: Does not have enough approvals]        
To /home/giteatest/gitea-repositories/user2/miomiomio.git
 ! [remote rejected] base -> master (pre-receive hook declined)
error: failed to push some refs to '/home/giteatest/gitea-repositories/user2/miomiomio.git'
```

Stack trace:
```
2020/01/12 20:36:05 ...ters/private/hook.go:145:HookPreReceive() [W] Forbidden: User 39 is not allowed push to protected branch master in 82473
8599184:user2/miomiomio and pr #1 is not ready to be merged: not allowed to merge [reason: Does not have enough approvals]
2020/01/12 20:36:05 routers/repo/pull.go:712:MergePullRequest() [E] Merge: git push: remote: Gitea: Not allowed to push to protected branch mas
ter and pr #68 is not ready to be merged: not allowed to merge [reason: Does not have enough approvals]
    To /home/giteatest/gitea-repositories/user2/miomiomio.git
     ! [remote rejected] base -> master (pre-receive hook declined)
    error: failed to push some refs to '/home/giteatest/gitea-repositories/user2/miomiomio.git'

    /home/guillep2k/src/code.gitea.io/gitea/routers/repo/pull.go:712 (0x1471838)
        MergePullRequest: ctx.ServerError("Merge", err)
    /home/guillep2k/go/src/reflect/value.go:460 (0x49e935)
        Value.call: call(frametype, fn, args, uint32(frametype.size), uint32(retOffset))
    /home/guillep2k/go/src/reflect/value.go:321 (0x49e0f3)
        Value.Call: return v.call("Call", in)
    /home/v/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:177 (0x9bc749)
        (*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil
    (etc)
```

User `39` is the owner of the repository and the user I'm logged in for the merge operation.

### PR DESCRIPTION
**Pull request by davidsvantesson**
Fix #9739 

- Admin shall be able to merge even if all checks are not ready (fix hook)
- If merge whitelist is explicitly set, admin should not be able to merge if not in whitelist. At least I didn't intended that change.

### CODE REVIEW

**Review by guillep2k — APPROVED** [review]

File: routers/private/hook.go [hunk 1]
```diff
@@ -139,19 +139,20 @@ func HookPreReceive(ctx *macaron.Context, opts private.HookOptions) {
 					})
 					return
 				}
-				// Manual merge only allowed if PR is ready (even if admin)
-				if err := pull_service.CheckPRReadyToMerge(pr); err != nil {
-					if models.IsErrNotAllowedToMerge(err) {
-						log.Warn("Forbidden: User %d is not allowed push to protected branch %s in %-v and pr #%d is not ready to be merged: %s", opts.UserID, branchName, repo, pr.Index, err.Error())
-						ctx.JSON(http.StatusForbidden, map[string]interface{}{
-							"err": fmt.Sprintf("Not allowed to push to protected branch %s and pr #%d is not ready to be merged: %s", branchName, opts.ProtectedBranchID, err.Error()),
+				if !perm.IsAdmin() {
```
Comment (lunny) [inline routers/private/hook.go:268]: Admin means repository admin permission but not site admin I think.

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/private/hook.go (hunk 1 above)
Comment (davidsvantesson) [reply routers/private/hook.go:268]: It is intended. Repo admin can bypass status checks and reviews for merging.

**Review by 6543 — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by davidsvantesson** [comment]
@zeripath Can you check my comments  in #9708? (sorry didn't have time to review before it was merged). I don't know how to resolve the conflicts on 136-137, line 136 seem to have no effect what I can see.

**Comment by davidsvantesson** [comment]
@zeripath I renamed `$notAllOk` to `$notAllOverrideableChecksOk`, I think it will be more clear. It should also be more clear if signed commits shall be overrideable in the future, so I don't think the term is needed there now.

**Review by zeripath — COMMENTED** [review]
Require signed commits here is being overrided by being the admin. If you actually intend that then you will need to change the code in pre-receive hook earlier where it checks that.

**Comment by davidsvantesson** [comment]
I removed the wrong location, fixed now.

**Comment by sapk** [comment]
Last change seems to fail on CI.

**Pull request merged**